task/gimlet-inspector/src/main.rs

```diff
@@ -0,0 +1,140 @@
+        // messages are small.
+        let mut rx_data_buf = [0u8; Request::MAX_SIZE + MAX_REQ_TRAILER_SIZE];
+        let mut tx_data_buf = [0u8; 128];
+
+        match net.recv_packet(
+            SOCKET,
+            LargePayloadBehavior::Discard,
+            &mut rx_data_buf,
+        ) {
+            Ok(mut meta) => {
+                let Ok((request, _trailer)) = hubpack::deserialize::<Request>(&rx_data_buf) else {
+                    // We ignore malformatted, truncated, etc. packets.
+                    continue;
+                };
+
+                match request {
+                    Request::V0(QueryV0::SequencerRegisters) => {
+                        let result = seq.read_fpga_regs();
+                        let (resp, trailer) = match result {
+                            Ok(regs) => (
+                                SequencerRegistersResponseV0::Success,
+                                Some(regs),
+                            ),
+                            Err(SeqError::ServerRestarted) => (
+                                SequencerRegistersResponseV0::SequencerTaskDead,
+                                None,
+                            ),
+                            Err(_) => {
+                                // The SeqError type represents a mashing
+                                // together of all possible errors for all
+                                // possible sequencer IPC operations. The only
+                                // one we _expect_ here is ReadRegsFailed.
+                                (SequencerRegistersResponseV0::SequencerReadRegsFailed, None)
+                            }
+                        };
+                        let mut len =
+                            hubpack::serialize(&mut tx_data_buf, &resp)
+                                .unwrap_lite();
+                        if let Some(t) = trailer {
+                            tx_data_buf[len..len + t.len()].copy_from_slice(&t);
+                            len += t.len();
+                        }
+                        meta.size = len as u32;
+                    }
+                }
+
+                // With the response packet prepared, we may need to attempt
+                // sending more than once.
+                loop {
+                    match net.send_packet(
+                        SOCKET,
+                        meta,
+                        &tx_data_buf[0..(meta.size as usize)],
+                    ) {
+                        Ok(()) => break,
+                        // If `net` just restarted, immediately retry our send.
+                        Err(SendError::ServerRestarted) => continue,
+                        // If our tx queue is full, wait for space. This is the
+                        // same notification we get for incoming packets, so we
+                        // might spuriously wake up due to an incoming packet
```

Comment:
for my own education: what happens to the incoming packet in this case?

Reply:
It hangs out in our RX queue until popped.

I actually just had to go read the netstack code to answer the anticipated follow-on question: so what determines when we get the next RX notification? The answer is, it depends on event flow in the netstack, and we are probably currently spamming any tasks that get into this state with RX notifications. I will file a bug.

Reply:
oh, interesting. seems like a good catch!